db/background_mgr_attachment_migration.go

```diff
@@ -0,0 +1,288 @@
+	migrationLoggingID := "Migration: " + a.MigrationID
+
+	persistClusterStatus := func() {
+		err := persistClusterStatusCallback(ctx)
+		if err != nil {
+			base.WarnfCtx(ctx, "[%s] Failed to persist cluster status on-demand for attachment migration operation: %v", migrationLoggingID, err)
+		}
+	}
+	defer persistClusterStatus()
+
+	var processFailure error
+	failProcess := func(err error, format string, args ...interface{}) bool {
+		processFailure = err
+		terminator.Close()
+		base.WarnfCtx(ctx, format, args...)
+		return false
+	}
+
+	callback := func(event sgbucket.FeedEvent) bool {
+		docID := string(event.Key)
+		collection := db.CollectionByID[event.CollectionID]
+		base.TracefCtx(ctx, base.KeyAll, "[%s] Received DCP event %d for doc %v", migrationLoggingID, event.Opcode, base.UD(docID))
+
+		// Don't want to process raw binary docs
+		// The binary check should suffice but for additional safety also check for empty bodies
+		if event.DataType == base.MemcachedDataTypeRaw || len(event.Value) == 0 {
+			return true
+		}
+
+		// We only want to process full docs. Not any sync docs.
+		if strings.HasPrefix(docID, base.SyncDocPrefix) {
+			return true
+		}
+
+		a.DocsProcessed.Add(1)
+		syncData, rawBody, rawXattrs, err := UnmarshalDocumentSyncDataFromFeed(event.Value, event.DataType, collection.userXattrKey(), false)
+		if err != nil {
+			failProcess(err, "error unmarshaling document %s: %v", base.UD(docID), err)
+		}
+
+		if rawXattrs[base.GlobalXattrName] != nil {
+			// global xattr already defined for this document, don't process
+			return true
+		}
+
+		if syncData == nil || syncData.Attachments == nil {
+			// no attachments to migrate
+			return true
+		}
+
+		collCtx := collection.AddCollectionContext(ctx)
+		collWithUser := &DatabaseCollectionWithUser{
+			DatabaseCollection: collection,
+		}
+		if event.DataType&base.MemcachedDataTypeXattr == 0 {
+			// no xattrs on dcp event we have inline sync xattr to migrate
+			opts := DefaultMutateInOpts()
+			existingDoc := &sgbucket.BucketDocument{
+				Cas:    event.Cas,
+				Body:   rawBody,
+				Expiry: event.Expiry,
+			}
+			_, _, err = collWithUser.migrateMetadata(ctx, docID, existingDoc, opts)
+			if err != nil {
+				failProcess(err, "error migrating document attachment metadata for doc: %s: %v", base.UD(docID), err)
+			} else {
+				a.DocsChanged.Add(1)
+			}
+		} else {
+			// normal xattr migration to take place
+			err = collWithUser.MigrateAttachmentMetadata(collCtx, docID, event.Cas, syncData)
+			if err != nil {
+				failProcess(err, "error migrating document attachment metadata for doc: %s: %v", base.UD(docID), err)
+			} else {
+				a.DocsChanged.Add(1)
+			}
+		}
+		return true
+	}
+
+	bucket, err := base.AsGocbV2Bucket(db.Bucket)
+	if err != nil {
+		return err
+	}
+
+	collectionIDs := getCollectionIDs(db)
+	dcpOptions := getMigrationDCPClientOptions(collectionIDs, db.Options.GroupID, db.MetadataKeys.DCPCheckpointPrefix(db.Options.GroupID))
```

Comment:
What happens in the case where we're resuming a previously running attachment migration job, but the set of collections on the database has changed since we last run?  We'd want to restart the DCP feed from zero, but I don't see that handled here.